pkg/depends_on.go

```diff
@@ -0,0 +1,35 @@
+package pkg
+
+import "fmt"
+
+func buildInstallOrder(svcs []Service) []string {
+	graph := Graph{nodes: []*Node{}}
+
+	nodeMap := map[string]*Node{}
+	for _, s := range svcs {
+		n := &Node{Name: s.Name}
+		nodeMap[s.Name] = n
+		graph.nodes = append(graph.nodes, n)
+	}
+
+	for _, s := range svcs {
+		for _, d := range s.DependsOn {
+			nodeMap[s.Name].Edges = append(nodeMap[s.Name].Edges, nodeMap[d])
+		}
+	}
+
+	resolved := &Graph{}
+	unresolved := &Graph{}
+	for _, g := range graph.nodes {
+		resolve(g, resolved, unresolved)
```

Comment:
This method seems to mix a few things that should be separated.  It looks like a private method that is specific to the Compose service handling, but it uses a generic and public `Graph` _and_ a private but generic method `resolve` and a private property of Graph.

I think that `resolve` should be public and either it is public in the package with the following signature
```go
Resolve(Node,Graph) (Graph,error)
```

or it is a mutating method on the Graph that sorts the nodes into their correct partial ordering

```go
// Sort puts the Graph Nodes into a partially sorted order based on their edges.  The
// implementation is based on https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
// this means that unrelated nodes (ie having no connecting edge) may be in the original order of addition to the Graph but any related nodes (ie having a connecting edge) will be sorted based on the edge direction.
func (g *Graph) Sort() error 
```

Reply:
Yes, it would be nice if the methods all belonged to the Graph struct. I'd be happy to take a PR or to work on polishing the API in a subsequent PR. My primary concern is fixing the underlying issue.

Reply:
The code would not be destructive (mutating), so if anything it would generate a new graph.

Reply:
This should be covered now in ->  https://github.com/openfaas/faasd/commit/5b92e7793dfebbd8e9b1700c2676cc5f56fad83e